### DISCUSSION TITLE
Support moving notes in hierarchies

### DISCUSSION THREAD

**Discussion by benjaminbauer** [artempyanykh/marksman#180] [Ideas]
I love the feature of lightweight [hierarchical notes](https://www.kevinslin.com/notes/3dd58f62-fee5-4f93-b9f1-b0f0f59a9b64/) in [dendron](https://github.com/dendronhq/dendron). Sadly dendron got abandoned my its maintainer and I am looking for a replacement.

In a nutshell: the filename of a note represents its hierarchy: e.g. `projects.private.pkm.marksman-evaluation.md`. Each hierarchy element may exist as an md file and have content itself and children. The big advantage: I do not have to create the whole hierarchy and I can freely move a note around in the hierarchy by renaming it.

I am thinking about writing a CLI tool for doing the hierarchy refactoring (one of the great enablers for such hierarchies). The workflow is roughly:

1. pick a note to move
2. offer fuzzy finding for the new parent in the hierarchy
3. confirm it (optional: preview)
4. Rename the note and its current children (could be configurable not to include the children
5. Rename the references to the renamed notes

Example:
I have the following notes:
1. `projects.private.pkm-evaluation.marksman.md`
2. `projects.private.pkm-evaluation.marksman.helix-intergration.md`
3. `projects.private.pkm-evaluation.dendron-evaluation.md`
4. `tools.pkm.md`

I want to move `marksman` to be a child of `tools.pkm`.
Target state:
1. `tools.pkm.marksman.md`
2. `tools.pkm.marksman.helix-intergration.md`
3. `projects.private.pkm-evaluation.dendron-evaluation.md`
4. `tools.pkm.md`

If I implement this as a separate CLI, I would have to copy or manually enter the note I want to move.

Is this a feature that could make it into marksman? Maybe as a (configurable?) refactoring option?

**Comment by artempyanykh** [comment]
Hi @benjaminbauer! Thanks for opening this.

In terms of hierarchical note you can already do it in marksman (albeit using path separator `/` rather than the dot `.`), e.g. `projects/private/pkm-evaluation/marksman.md`

Moving notes in the hierarchy could be done with rename refactoring. LSP already has hooks such as `willRenameFile`. This can be used to fix up all the links, e.g.
1. You rename the root note `projects/private/pkm-evaluation/marksman.md` to `tools/pkm/marksman.md`; editor sends a `willRenameFile` request to Marksman; Marksman returns edits that your editor applies to fix up the references.
2. Your rename the folder `projects/private/pkm-evaluation/` to `tooks/pkm/marksman/`. Same as above happens for notes inside this folder.

You'd need to do 2 renames rather than one but at least this can be done in a standard LSP way without a bespoke integration.

Marksman already has support for rename refactoring for headings and references, and I think it can be more or less easily extended to file names.

Does this help?

**Reply by benjaminbauer** [reply]
Thanks for the instant reply :) 
Is there documentation for the path feature? Do I understand it correctly, that this feature depends on folders rather than the file name? If so, how is the transition handled when I decide that e.g. `pkm-evaluation` shall have content, i.e. be an md file?

**Comment by artempyanykh** [comment]
Relevant issue #153

**Comment by ghost** [comment]
I would also like to see this. I have multiple files with the same syntax. Marksman does not recognize it. For example, `animals.cat` is converted to `animalscat`, which makes difficult to read. Moreover using "/" does not solve the problem because the intention with hierarchical filenames is avoiding the use of folders to create the hierarchy.
